zeus/monitor.py

```diff
@@ -0,0 +1,230 @@
+# Copyright (C) 2022 Jae-Won Chung <jwnchung@umich.edu>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Helpers for using the Zeus monitor inside training scripts for energy profiling."""
+
+from __future__ import annotations
+
+import time
+import atexit
+import signal
+import tempfile
+import subprocess
+from typing import Generator
+from contextlib import contextmanager
+
+import torch
+
+from zeus.analyze import energy as compute_energy
+
+
+class ZeusMonitorContext:
+    """Monitors the energy and time consumption inside a training loop.
```

Comment:
I'm not very sure about the exact objective of this class. Based on the code, it will be one `ZeusMonitorContext` per epoch. Within one epoch, it will keep doing (skip + profile). Do we need to keep track of some values over one epoch? For example `avg_energy`? Or it's just for logging purpose

Reply:
The integration example does show that there is one `zeus_ctx` per epoch, but it doesn't have to be that way. The user can create a global `zeus_ctx` object and reuse it throughout training. Based on the use case, the user can just profile one cycle (e.g., 10 skips, 20 profiles), and report numbers, or if the user is doing something like JIT profiling, can use `zeus_ctx` repeatedly on multiple profiling windows.

Reply:
In `dataloader`, we start a monitor at the beginning of each epoch and kill it at the end of the epoch. What's the reason for this specific design? Would it be the same case here? Or the monitor here can across epochs?

Reply:
I see. That makes sense 👍